### ISSUE TITLE
Configure flux-specific RBAC for kubeapps-apis

### ISSUE THREAD

**Issue by absoludity** [kubeapps/kubeapps#3494]
### Description:

Currently the flux plugin requires either running the kubeapps-apis server with the unsafe, dev-only RBAC for testing, or manually configuring extra RBAC for the service, so that the linked service account has the access control required by the plugin.

The extra RBAC is required because the flux plugin starts its own background processes that watch resources in the cluster asynchronously (outside of the user interactions, which we usually use to limit required RBAC).

This RBAC should be moved to a separate file such as `chart/kubeapps/templates/kubeappsapis/rbac_fluxv2.yaml` that is only installed if `enabledPlugins` contains `fluxv2`, defining a cluster role and role binding for the service account.

It'd be great to minimize the RBAC required too, of course (currently verb: * etc.)

**Comment by gfichtenholt** [comment]
First, I'd like to make sure I understand the problem correctly. I think the gist of what you're saying is we should not need to use "--set kubeappsapis.unsafeUseDemoSA=true" when doing helm install of kubeapps to be able use flux plug-in. Today we have to. If so, sounds reasonable. Hopefully I got that right :-)

Secondly, looks like there is a little more to do than that is described above. Based on looking at current code, it looks like the service account "kubeapps-internal-kubeappsapis" is not even created unless one uses "--set kubeappsapis.unsafeUseDemoSA=true" flag. Is that intentional? If so, this is a little more than adding some RBAC to an existing service account. I think I'd need to:
1) create a service account along with clusterrole and binding. Would that be a flux plug-in specific service account? 
2) if we don't want to use any service account in production for running kubeapps-api deployment (is that true?), then I need to programmatically run the flux plug-in code as (1)

Or maybe the service account "kubeapps-internal-kubeappsapis" was really meant to be created regardless of "--set kubeappsapis.unsafeUseDemoSA=true" flag and kubeappsapis deployment is meant to be using this service account? That would follow the same pattern with apprepository controller and kubeops? 

In short, I am a little confused and need clarification before I proceed. Thanks

**Comment by antgamdia** [comment]
My two cents here:

> I think the gist of what you're saying is we should not need to use "--set kubeappsapis.unsafeUseDemoSA=true" when doing helm install of kubeapps to be able use flux plug-in. 

Yep, I also think so. I'm confident we shouldn't require using any `unsafeXXX` option for deploying Kubeapps whatsoever.

>  it looks like the service account "kubeapps-internal-kubeappsapis" is not even created unless one uses "--set kubeappsapis.unsafeUseDemoSA=true" flag

Yes, that is correct. The SA is not created unless this flag is passed. The motivation behind this is simply we thought we wouldn't require any SA at all when running in production because every action is performed on a user's behalf. However, the use case in the flux plugin is slightly different in that we need some sort of background process.

Both approaches you mention will solve the issue, of course. Whereas following the same pattern as in the kubeops service (a privileged single SA) is an option, I don't think we should go this way. Following the "zero trust" model, I think we should grant as few permissions as possible, so a plugin-specific SA is a better approach IMHO.

So, what I would do is: i) removing the unsafeXXX flags from the code; ii) replacing the current SA with a plugin-specific one, and edit the condition to marking the plugin "flux" as enabled; iii) using this SA in the flux plugin code.

Don't know if Michael thinks otherwise, though

**Comment by gfichtenholt** [comment]
Off the top of my head, can't say I fully agree. True, Flux plug-in is somewhat of a "[controller](https://kubernetes.io/docs/concepts/architecture/controller/)" in that it does things in the background. In fact, in the beginning, I planned to write a proper k8s controller for flux, but Michael suggested I just do what I need in a goroutine, so that's where we are today. Still, in my mind, it's not too far fetched to think of flux plugin as a controller. Now, the plug-in is loaded into kubeapps-apis-server as a module, which transitively sort of makes kubeapps-apis-server a controller. Now the usual pattern for controllers is to run as some service account. For example, in flux itself:
```
$ kubectl get serviceaccounts -n flux-system
NAME                          SECRETS   AGE
default                       1         161m
helm-controller               1         161m
image-automation-controller   1         161m
image-reflector-controller    1         161m
kustomize-controller          1         161m
notification-controller       1         161m
source-controller             1         161m
```
And corresponding deployments run with these accounts. 
```
$ kubectl get deployment helm-controller -n flux-system -o yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    ...
  name: helm-controller
  namespace: flux-system
  resourceVersion: "6425"
  uid: e787ed48-8457-47ab-b7e3-e689d9cd5aa4
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: helm-controller
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        prometheus.io/port: "8080"
        prometheus.io/scrape: "true"
      creationTimestamp: null
      labels:
        app: helm-controller
    spec:
      containers:
      - args:
        - --events-addr=http://notification-controller/
        - --watch-all-namespaces
        - --log-level=info
        - --log-encoding=json
        - --enable-leader-election
        env:
        - name: RUNTIME_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        image: fluxcd/helm-controller:v0.11.2
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: healthz
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: manager
        ports:
        - containerPort: 8080
          name: http-prom
          protocol: TCP
        - containerPort: 9440
          name: healthz
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /readyz
            port: healthz
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 1Gi
          requests:
            cpu: 100m
            memory: 64Mi
        securityContext:
          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /tmp
          name: temp
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: helm-controller
      serviceAccountName: helm-controller
      terminationGracePeriodSeconds: 600
      ...
```
So I guess what I am saying an approach that seems cleaner to me as to have a kubeapps-apis-server run under a specific service account, BUT assign different RBACs to it depending on which plug-ins are enabled. Without flux enabled these RBACs will be very minimal, with flux enabled they would be more. So, focus on declarative (easier to change/maintain) way, rather that programmatic (harder to change/maintain). Otherwise, the burden on plug-ins with similar issues seems kind of big. Having said that, I do not insist on my interpretation of things, I could be convinced otherwise

**Comment by absoludity** [comment]
> So I guess what I am saying is an approach that seems cleaner to me is to have a kubeapps-apis-server run under a specific service account,

+1

> BUT assign different RBACs to it depending on which plug-ins are enabled. Without flux enabled these RBACs will be very minimal,

Yes, without the flux plugin, there will either be no role bound to the service account currently I think? (in the future, we will indeed need a minimal role bound so that the service can list namespaces when configured, like kubeops currently does, but that's not yet part of the apis server).

**Comment by gfichtenholt** [comment]
sounds good to me. I will work on it tomorrow+
